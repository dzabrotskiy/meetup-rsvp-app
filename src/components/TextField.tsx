import { ForwardedRef, forwardRef, Fragment } from 'react';
import {
  StyleProp,
  StyleSheet,
  TextInput,
  TextInputProps,
  TextStyle,
} from 'react-native';

import { Text, useThemeColor, View } from './Themed';
import { Spacer } from './Spacer';

export type Props = {
  placeholder?: string;
  label?: string;
  status?: 'default' | 'error';
  description?: string;
  style?: StyleProp<TextStyle>;
  value: TextInputProps['value'];
  onChange: TextInputProps['onChangeText'];
  onBlur?: TextInputProps['onBlur'];
  keyboardType?: TextInputProps['keyboardType'];
  multiline?: TextInputProps['multiline'];
  maxLength?: TextInputProps['maxLength'];
};

export const TextField = forwardRef(function TextField(
  {
    value,
    onChange,
    label,
    placeholder,
    keyboardType,
    onBlur,
    multiline,
    maxLength,
    style,
    status,
    description,
  }: Props,
  ref: ForwardedRef<TextInput>
) {
  const backgroundColor = useThemeColor('secondary');
  const textColor = useThemeColor('text');

  return (
    <View style={styles.container}>
      {label && <Text style={styles.label}>{label}</Text>}
      <Spacer height={6} />
      <TextInput
        ref={ref}
        value={value}
        onChangeText={onChange}
        placeholder={placeholder}
        style={[
          { backgroundColor, color: textColor },
          styles.input,
          style,
          status === 'error' && styles.inputError,
        ]}
        keyboardType={keyboardType}
        onBlur={onBlur}
        multiline={multiline}
        maxLength={maxLength}
      />
      {description && (
        <Fragment>
          <Spacer height={3} />
          <Text
            style={[
              styles.description,
              status === 'error' && styles.descriptionError,
            ]}
          >
            {description}
          </Text>
        </Fragment>
      )}
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'transparent',
  },
  label: {
    fontSize: 13,
    marginLeft: 10,
  },
  input: {
    height: 45,
    borderRadius: 10,
    fontSize: 19,
    paddingLeft: 10,
  },
  inputError: {
    borderWidth: 1,
    borderColor: '#f00',
  },
  description: {
    fontSize: 11,
    marginLeft: 10,
  },
  descriptionError: {
    color: '#f00',
  },
});
