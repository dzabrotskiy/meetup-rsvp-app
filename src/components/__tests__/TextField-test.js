import renderer from 'react-test-renderer';
import * as React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import '@testing-library/jest-native/extend-expect'

import { TextField } from '../TextField';

describe('TextField', () => {
  it('should match snapshot', () => {
    const tree = renderer
      .create(<TextField value="value" onChange={() => {}} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call onChange', () => {
    const onChange = jest.fn();
    const { getByPlaceholderText } = render(
      <TextField value="" onChange={onChange} placeholder="Placeholder" />
    );

    const textInput = getByPlaceholderText('Placeholder');
    fireEvent.changeText(textInput, 'text');

    expect(onChange).toBeCalledWith('text');
  });

  it('should have correct label', () => {
    const onChange = jest.fn();
    const { getByText } = render(
      <TextField
        value=""
        onChange={onChange}
        placeholder="Placeholder"
        label="Label"
      />
    );

    expect(getByText('Label')).toBeDefined();
  });

  it('should render error state', () => {
    const onChange = jest.fn();
    const { getByPlaceholderText } = render(
      <TextField
        value=""
        onChange={onChange}
        placeholder="Placeholder"
        label="Label"
        status="error"
      />
    );

    const textInput = getByPlaceholderText('Placeholder');

    expect(textInput).toHaveStyle({
      borderWidth: 1,
      borderColor: '#f00'
    })
  })
});
