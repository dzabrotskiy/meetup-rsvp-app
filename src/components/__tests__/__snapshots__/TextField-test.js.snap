// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should match snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "backgroundColor": "transparent",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "height": 6,
          },
          {
            "backgroundColor": "transparent",
          },
        ],
      ]
    }
  />
  <TextInput
    onChangeText={[Function]}
    style={
      [
        {
          "backgroundColor": "#e0e0e0",
          "color": "#000",
        },
        {
          "borderRadius": 10,
          "fontSize": 19,
          "height": 45,
          "paddingLeft": 10,
        },
        undefined,
        false,
      ]
    }
    value="value"
  />
</View>
`;
